plugins/fileservices/fileservices.cpp

```diff
@@ -2726,70 +2726,152 @@ FILESERVICES_API char *  FILESERVICES_CALL fsfResolveHostName(const char *hostna
+
+    CDfsLogicalFileName dlfn;
+    if (plane)
+    {
+        //allow relative path if a plane name is specified.
+        if (isAbsolutePath(path))
+        {
+            const char *relativePath = getRelativePath(path,plane->queryProp("@prefix"));
+            if (nullptr == relativePath)
+                throw makeStringExceptionV(-1,"Invalid plane path %s.",path);
+            path = relativePath;
+        }
+        dlfn.setPlaneExternal(plane->queryProp("@name"),path);
+    }
+    else
+        dlfn.setExternal(host,path); //for backward compatibility
+    dlfn.getExternalFilename(rfn);
+
+    Owned<IFile> f = createIFile(rfn);
+    if (rd && !f->exists())
+        throw makeStringExceptionV(-1,"File %s not found",path);
+
+    if (f->exists() && (f->isDirectory()==fileBool::foundYes))
+        checkExternalFileRights(ctx,dlfn.get(),rd,wr); //check if the folder can be read/writen.
+    else
+    {
+        StringBuffer scopes;
+        checkExternalFileRights(ctx,dlfn.getScopes(scopes),rd,wr);
+    }
+}
+
+static IPropertyTree *checkPlaneOrHost(const char *planeName,const char *host)
+{
+    if (!isEmptyString(planeName))
+    {
+        Owned<IPropertyTree> plane = getDropZonePlane(planeName); //Only support DropZone for now.
+        if (!plane)
+            throw makeStringExceptionV(-1,"DropZone %s not found.",planeName);
+        return plane.getClear();
+    }
+    SocketEndpoint ep(host);
     if (ep.isNull())
-        throw MakeStringException(-1,"fsMoveExternalFile: Cannot resolve location %s",location);
-    CDfsLogicalFileName from;
-    from.setExternal(location,frompath);
-    CDfsLogicalFileName to;
-    to.setExternal(location,topath);
-    checkExternalFileRights(ctx,from,true,true);
-    checkExternalFileRights(ctx,to,false,true);
-    RemoteFilename fromrfn;
-    fromrfn.setPath(ep,frompath);
-    RemoteFilename torfn;
-    torfn.setPath(ep,topath);
-    Owned<IFile> fileto = createIFile(torfn);
-    if (fileto->exists())
-        throw MakeStringException(-1,"fsMoveExternalFile: Destination %s already exists", topath);
-    fileto.clear();
-    Owned<IFile> file = createIFile(fromrfn);
-    file->move(topath);
+        throw makeStringExceptionV(-1,"Cannot resolve host %s",host);
+    return nullptr;
+}
+
+static void implementMoveExternalFile(ICodeContext *ctx,const char *location,
+    const char *frompath,const char *topath,const char *planename)
+{
     StringBuffer s("MoveExternalFile ('");
-    s.append(location).append(',').append(frompath).append(',').append(topath).append(") done");
-    WUmessage(ctx,SeverityInformation,NULL,s.str());
-    AuditMessage(ctx,"MoveExternalFile",frompath,topath);
+    try
+    {
+        Owned<IPropertyTree> plane = checkPlaneOrHost(planename,location);
+        RemoteFilename fromrfn,torfn;
+        checkExternalFilePath(ctx,plane,location,frompath,true,true,fromrfn);
+        checkExternalFilePath(ctx,plane,location,topath,false,true,torfn);
```

Comment:
haven't thought about too deeply, but feels like dropzone vs paths within dropzones checking + perm. checking functionality should already be covered by existing implemented functionality e.g. validateDZPathScopePermsAndLegacyPhysicalPerms etc.

Plus, see other comment re. if if unknown whether file or dir, assume file, and check user has perms to access parent dir. 1st, before returning checking returning error that file doesn't exist.

Reply:
> validateDZPathScopePermsAndLegacyPhysicalPerms

@wangkx I realize methods like validateDZPathScopePermsAndLegacyPhysicalPerms are currently in SMCLib and use IEspContext's and therefore inaccessible to the fileservices plugin.

But if this is a common functional need, that currently requires very similar functionality to be duplicated, it suggests it should be commoned up somewhere, where it is accessible by both.

It may be that it is best for this PR to leave it a duplicated functionality, but open a JIRA and revisit to refactor to avoid the duplication/future confusion.

Reply:
I opened https://track.hpccsystems.com/browse/HPCC-31280 to refactor to avoid the duplication/future confusion.